src/v/iceberg/avro_utils.h

```diff
@@ -31,7 +33,12 @@ class avro_iobuf_ostream : public avro::OutputStream {
     // space.
     bool next(uint8_t** data, size_t* len) final {
         if (available_ == 0) {
-            buf_->append(ss::temporary_buffer<char>{chunk_size_});
+            // NOTE: it is critical to add the buffer to a separate iobuf first
+            // and then add that iobuf's fragments, as adding a buffer directly
+            // to `buf_` may end up packing the fragments together.
+            iobuf new_frag;
+            new_frag.append(ss::temporary_buffer<char>{chunk_size_});
+            buf_->append_fragments(std::move(new_frag));
```

Comment:
> This led to exceptions being thrown when serializing
larger Avro files.

like what? this is a concerning result because generally the fragmentation of data within an iobuf should not affect correctness provided that users of the iobuf are aware of this.

Reply:
Ah, this statement isn't quite true -- it was that we would serialize incorrectly, and we'd see exceptions when reading them.

>the fragmentation of data within an iobuf should not affect correctness provided that users of the iobuf are aware of this

I think it's the case here because we're working around the iobuf interface to pass contiguous buffers to the DataFileWriter, which just gets a uint8_t* and a size.

Reply:
> it is critical to add the buffer to a separate iobuf first
> and then add that iobuf's fragments, as adding a buffer directly
> to `buf_` may end up packing the fragments together.

i'm still not clear on what's going on. this comment suggests to me that that something may be broken, or very fragile.

is this about pointer stability of the data in the temporary buffers? if so (what follows assume this), iobuf doesn't really make any explicit guarantees about this. it could offer those guarantees, and in practice i think it does for the interface you are using. but AFAIK it's not written in stone, so this could end up biting us later on.

what if you held on to the raw temporary buffers explicitly, and then built the iobuf on demand later when pointer stability was no longer required? or maybe we need to codify some guarantees in iobuf interface?

Reply:
>is this about pointer stability of the data in the temporary buffers?

Yea this is correct, fair point about iobuf's contract to callers.

>what if you held on to the raw temporary buffers explicitly, and then built the iobuf on demand later when pointer stability was no longer required? or maybe we need to codify some guarantees in iobuf interface?

This seems reasonable. Mind if I do it in a follow-up change?

Reply:
within what scope does pointer stability need to be guaranteed? does it need to be maintained for the life of the iobuf, or for a shorter scope?

Reply:
We chatted a bit about this offline.

The pointers are required to be stable for at least the duration during which the DataFileWriter writes to the stream. In practice, I think this means for the entire lifespan of the DataFileWriter.

We also discussed a couple approaches to making this less fragile. I've implemented one: where callers are now required to manage the ss::temporary_buffers, and pass them into an iobuf once the writer is finished. This is added in the latest push.